### ISSUE TITLE
[BUG] rdiff-backup 2.0.0 has problems when the backup target is a btrfs subvolume

### ISSUE THREAD

**Issue by MaxRower** [rdiff-backup/rdiff-backup#798]
## Bug summary

Backing up to the subvolume directory is fine, but "rdiff-backup -l target", "rdiff-backup --list-increment-sizes target" or "rdiff-backup --remove-older-than 180B target" does not recognize it correctly and reports 
"Fatal Error: Bad directory xxxx.
It doesn't appear to be an rdiff-backup destination dir."

Creating a directory below the subvolume directory and moving all file there does help.

## Version, Python, Operating System

Ubuntu 20.04.5 LTS (GNU/Linux 5.4.0-132-generic x86_64

Call `rdiff-backup info` (version >= 2.1) or `rdiff-backup -v9` and replace the following line with the output, repeat for each environment impacted:

```yaml
2022-11-30 12:50:33.018023 +0100  <CLIENT-3110>  Using rdiff-backup version 2.0.0
2022-11-30 12:50:33.018110 +0100  <CLIENT-3110>         with cpython /usr/bin/python3 version 3.8.10
2022-11-30 12:50:33.028424 +0100  <CLIENT-3110>         on Linux-5.4.0-132-generic-x86_64-with-glibc2.29, fs encoding utf-8
2022-11-30 12:50:33.028528 +0100  <CLIENT-3110>  Fatal Error: No arguments given
See the rdiff-backup manual page for more information.

```

## rdiff-backup call

How did you call rdiff-backup, with which parameters:

```
rdiff-backup -l targetdir
```

## What happened and what did you expect?

rdiff-backup does not recognize the subvolume/directory correctly and reports the following error:
"Fatal Error: Bad directory xxxx.
It doesn't appear to be an rdiff-backup destination dir."

## More information

root@backup:/srv/backup/containers# btrfs subvolume list /srv/ | grep backup/containers/home-fkg
ID 8594 gen 144805 top level 5 path backup/containers/home-fkg
root@backup:/srv/backup/containers# ls home-fkg/
backup.yaml  rdiff-backup-data  rootfs
root@backup:/srv/backup/containers# rdiff-backup -l home-fkg/
Fatal Error: Bad directory home-fkg.
It doesn't appear to be an rdiff-backup destination dir.
root@backup:/srv/backup/containers# mkdir home-fkg/dir
root@backup:/srv/backup/containers# mv home-fkg/* home-fkg/dir
mv: cannot move 'home-fkg/dir' to a subdirectory of itself, 'home-fkg/dir/dir'
root@backup:/srv/backup/containers# ls home-fkg/dir/
backup.yaml  rdiff-backup-data  rootfs
root@backup:/srv/backup/containers# rdiff-backup -l home-fkg/dir/
Found 0 increments:
Current mirror: Wed Nov 30 10:30:38 2022

**Comment by ericzolf** [comment]
Can you please repeat with 2.0.5, or even better with the latest beta version from this site? 2.2 will be soon released.

**Comment by MaxRower** [comment]
> Can you please repeat with 2.0.5, or even better with the latest beta version from this site? 2.2 will be soon released.

I did a quick container install with Ubuntu 22.10 and it's rdiff-backup 2.0.5, created a subvolume inside the container (that I could not list with btrfs subvolume list inside the container, which is hopefully not the reason why it does work here), and made two rdiff-backups to that subvolume/directory. rdiff-backup -l, --list-increment-sizes and --remove-older-than did work without errors. So that bug seems to be fixed in later versions than 2.0.0. Too bad, Ubuntu 20.04 does not provide a newer version, I won't be able to upgrade any time soon.

**Comment by ikus060** [comment]
@MaxRower 
For Minarca and Rdiffweb, I need to compile recent version of rdiff-backup. That include version 2.0.5 for Ubuntu Kinetic (22.10)

I don't have a clear procedure, but quickly, that should be as follow:

```
apt install lsb-release
curl -L https://www.ikus-soft.com/archive/rdiffweb/public.key | gpg --dearmor > /usr/share/keyrings/rdiffweb-keyring.gpg
echo "deb [signed-by=/usr/share/keyrings/rdiffweb-keyring.gpg] https://nexus.ikus-soft.com/repository/apt-release-$(lsb_release -sc)/ $(lsb_release -sc) main" > /etc/apt/sources.list.d/rdiffweb.list
apt update
apt install rdiff-backup2
```

That will install rdiff-backup 2.0.5 side-by-side with the installation provided by Ubuntu. The binaries are installed under `/opt`.

**Comment by MaxRower** [comment]
I found the following PPA https://launchpad.net/~rdiff-backup/+archive/ubuntu/rdiff-backup-backports?field.series_filter=focal
Installed it, now I'm runnung 2.0.5, but the error still occurs.

**Comment by ericzolf** [comment]
OK, then let's re-open.

Would you mind trying the latest beta version, I've just documented how to do it simply at https://github.com/rdiff-backup/rdiff-backup/blob/ericzolf-doc-virtualenv/docs/DEVELOP.adoc#install-for-test-and-development (version to use is as in the 2nd installation example i.e. `2.1.3b3`)?

I don't have btrfs in my experience or on my computers, so I'll need a bit of time to set it up, would be nice to know that the error is still there before I invest this effort.

**Comment by MaxRower** [comment]
Just noticed, every subvolume on the file system has d--x------ permissions (seems to be the standard), maybe this is the cause?

**Comment by MaxRower** [comment]
Same behaviour, when I set the permissions of a non-subvolume-target to 100. So it's only the permissions, not subvolume vs. normal directory.

**Comment by ericzolf** [comment]
Are you actually really root or only a pseudo root in a container? I'm asking this because on my laptop, calling rdiff-backup as root on such a directory, everything works as it should, whereas a normal user fails miserably (as it should).

I even completely remove the access rights, and that doesn't interest root at all, which is what I'd expect from root, it simply works in all conditions. So IMHO:

1. either your root isn't really root
2. or there are further access rights in the game which restrict what root can do:
    1. either via SELinux or similar (FACL?)
    2. or through the file system itself

I doubt that the failure is due to rdiff-backup at this stage.

**Comment by ericzolf** [comment]
I created a btrfs volume and sub-volume, did a backup `rdiff-backup -v5 backup /var/lib/containers/storage /mnt/sub_vol/`, the access rights were `u=rwx`, I removed the access rights `chmod -R a-rw,go-x sub_vol/`, and still `rdiff-backup list increments sub_vol/` resp. `rdiff-backup -l sub_vol/` doesn't complain. That's with the latest development version. Interestingly enough the old 2.0.5 version does indeed fail on the same directory.

So, two conclusions and I'll close this issue:

1. rdiff-backup 2.0.5 was indeed acting strange
2. rdiff-backup 2.2.0 will have fixed the issue (to be released before end of the year)
